mllib-local/pom.xml

```diff
@@ -75,48 +75,12 @@
       <type>test-jar</type>
       <scope>test</scope>
     </dependency>
+
+    <dependency>
+      <groupId>dev.ludovic.netlib</groupId>
+      <artifactId>blas</artifactId>
+    </dependency>
   </dependencies>
-  <profiles>
-    <profile>
-      <id>netlib-lgpl</id>
```

Comment:
I think we need to keep this profile, even if empty, to avoid failing builds that specify `-Pnetlib-lgpl` (think of it as deprecated).

But, something needs to pull in the `all` artifact, or else building Spark with this profile doesn't include native libs for acceleration, right? and those need to be behind a profile. I'd not expect this profile goes away.

Maybe you can show the relevant part of `mvn dependency:tree` or the result of `dev/test-dependencies.sh --replace-manifest` so we can be clear about what packages are now included at the moment. I want to make sure the default doesn't include LGPL code.

Reply:
Is it enough to have this profile in `pom.xml` ([like so](https://github.com/apache/spark/pull/32253/files/#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R3489-R3499)) or does it need to be in this `mllib-local/pom.xml`?

The results from `dev/test-dependencies.sh --replace-manifest` __without__ `-Pnetlib-lgpl` are at https://gist.github.com/luhenry/ceb64c52ad63c69ed0ef2aa205ddde25. The only added dependencies are the blas, lapack, and arpack packages from `dev.ludovic.netlib`.

Reply:
I believe we'll need the profile in both places - parent is more about "declaration" and child is more about actually using it.

The diff looks OK, but raises the question, what's in `arpack`, `lapack`, etc? does that build in native code? I'd have to go check but looks like LAPACK et al are BSD licensed, which is OK.

But it looks like nothing references the `all` artifact from netlib, so not sure what would build in the native code to actually do acceleration, even with the profile? I'm probably missing something.

Reply:
> I believe we'll need the profile in both places - parent is more about "declaration" and child is more about actually using it.

Ok, let me revert it to how it's in `master`.

> The diff looks OK, but raises the question, what's in arpack, lapack, etc? does that build in native code? I'd have to go check but looks like LAPACK et al are BSD licensed, which is OK.

These only rely on`com.github.fommil.netlib:core` just like `blas`. These do not include any native code, and only provide a wrapper for `com.github.fommil.netlib-java` for now. In the future, I want to provide the same native wrappers as for `blas` but, similarly, without including any actual native code, just the JNI-based wrappers.

> But it looks like nothing references the all artifact from netlib, so not sure what would build in the native code to actually do acceleration, even with the profile? I'm probably missing something.

I am not sure what you don't understand. What do you mean in "__what__ would build in the native code"?

If I understand correctly your question, the native code is wrapped by `com.github.fommil.netlib` for JDK8 and JDK 11. For JDK16, `dev.ludovic.netlib` does provide a wrapper using the Foreign Linker API, and falls back to using `com.github.fommil.netlib` if the Foreign Linker API is not enabled. In the future, I do want to add a JNI-based wrapper in `dev.ludovic.netlib` directly so that I don't have to rely on `com.github.fommil.netlib` itself.

What I am not sure I understand in the licensing issue is what is licensed as GPL/LGPL? Looking at https://mvnrepository.com/artifact/com.github.fommil.netlib/all/1.1.2, it all seems to be licensed as BSD. So is it the Fortran/native code that's shipped with it which is licensed as GPL and which is an issue?

Thank you again for helping me navigating this problem!

Reply:
First some background - I think the issue with licensing comes from the submodules that are pulled in by `all`, like: https://github.com/fommil/netlib-java/blob/master/native_ref/xbuilds/linux-x86_64/pom.xml netlib-java generates and compiles native code. So things like https://mvnrepository.com/artifact/com.github.fommil.netlib/netlib-native_system-linux-x86_64/1.1 ship a `.so` and I believe it's that code or what it links to that is LGPL. I've honestly forgotten the details but seemed pretty clear (it may be, at least, libgfortran). At least, I recall going over this with a fine-toothed comb with Cloudera's legal department and they agreed, FWIW.

Anyway, that's the reason behind all this bother with profiles. The wrapper in netlib-java is OK to include in distributions and compile against. But adding in the dependencies that bring in native code is not. Hence it's under a profile that downstream users can turn on for their own builds if desired.

So, my question is, where is the native code coming in here to do the acceleration? I understand Java 16+ has it built in, that's the whole point of your change, but we still want to let people flip on dependencies that will build in bindings to OpenBLAS, etc. Where would that be now? because I see you're just depending on netlib-java `core`, which doesn't have that. I'm asking if that's what `blas` is, etc, but not sure that's it either.

Reply:
> First some background - I think the issue with licensing comes from the submodules that are pulled in by all, like: https://github.com/fommil/netlib-java/blob/master/native_ref/xbuilds/linux-x86_64/pom.xml netlib-java generates and compiles native code. So things like https://mvnrepository.com/artifact/com.github.fommil.netlib/netlib-native_system-linux-x86_64/1.1 ship a .so and I believe it's that code or what it links to that is LGPL. I've honestly forgotten the details but seemed pretty clear (it may be, at least, libgfortran). At least, I recall going over this with a fine-toothed comb with Cloudera's legal department and they agreed, FWIW.

I can confirm that both `netlib_ref` and `netlib_system` do link to `libgfortran` and which is licensed with LGPL.  See [[1]](https://github.com/fommil/netlib-java/blob/master/native_ref/xbuilds/linux-x86_64/pom.xml#L66) and [[2]](https://github.com/fommil/netlib-java/blob/master/native_system/xbuilds/linux-x86_64/pom.xml#L77).

> Anyway, that's the reason behind all this bother with profiles. The wrapper in netlib-java is OK to include in distributions and compile against. But adding in the dependencies that bring in native code is not. Hence it's under a profile that downstream users can turn on for their own builds if desired.

So if I understand correctly, the limitation is that any code that statically or dynamically links to any LGPL library is out of the picture for _distribution_ with Spark. And if any downstream user wants to do it, they can but the Open Source Spark project won't. Is that correct?

> So, my question is, where is the native code coming in here to do the acceleration?

The main difference that `dev.ludovic.netlib` brings compared to `com.github.fommil.netlib:native_ref` or even `com.github.fommil.netlib:native_system` is that it doesn't link explicitely to any native library (OpenBLAS, MKL, libgfortran) at build-time (there is no `-l...` linker flag). Instead, it probes at run-time (via `dlopen`/`dlsym`) for a library which name is `libblas.so` (or anything else specified by the user) and which exposes certain symbols (`cblas_dgemm`, `cblas_ddot`, for example). It doesn't distribute any such library and it doesn't link to any such library (statically or dynamically) or any of their dependencies.

If the user wants to use OpenBLAS or Intel MKL, or any other BLAS-like library for that matter, it's up to them to install it on their system. Then, either the library is named `libblas.so` (or there is such a symbolic filesystem link) and is accessible via LD_LIBRARY_PATH, or the user specifies the name with `-Ddev.ludovic.netlib.blas.nativeLib` or full path with `-Ddev.ludovic.netlib.blas.nativeLibPath`.

Overall `dev.ludovic.netlib` doesn't link to, distribute, nor even favor any specific native BLAS implementation. So, by extension, Spark doesn't link to, distribute, nor favor any specific native BLAS implementation or external library (like libgfortran) with this patch.

> I understand Java 16+ has it built in, that's the whole point of your change, but we still want to let people flip on dependencies that will build in bindings to OpenBLAS, etc. Where would that be now?

`dev.ludovic.netlib.blas.ForeignBLAS` does load an arbitrary `libblas.so` library and probes for symbols like `cblas_dgemm`, `cblas_ddot`, etc. It's using `dlopen`/`dlsym` in the JVM directly.

> because I see you're just depending on netlib-java core, which doesn't have that. I'm asking if that's what blas is, etc, but not sure that's it either.

`blas` contains two things:
1. the java implementations of the BLAS libraries with fallbacks to F2jBLAS (that's VectorizedBLAS for example)
2. the bindings for native BLAS libraries (with the Foreign Linker API for now, and JNI-based bindings in the future).

The contention point is in `2. the binding for native BLAS libraries` AFAIU. If it's a blocker, I should be able to break it down into `blas` and `blas-native` packages for example.

Reply:
That's right. The ASF needs to release artifacts that are as or no more restrictive than ALv2, and that ends mostly meaning: no binary artifacts with copy-left or field-use-of-restricted licenses like the GPL family. It does turn on releasing a copy of code, or in the case of GPL, more about linking to it.

The typical policy is that it's OK to release software that can merely make use of such libraries at runtime (without actually distributing them directly) as long as it doesn't substantially depend on their presence. I believe that dynamic linking in the way you describe is OK - just like having an SPI in JVM code that may be provided by some other GPL code at the user's runtime.

My main goal is to preserve current behavior. Right now if someone has, say, MKL on their native lib path for the JVM, and built with this alternate profile, it'd be accelerated. If you're saying that still works, but would not require this separate build profile because of the different loading strategy, that's an improvement.

If nothing that ends up in transitive dependencies can be called statically linked to, say, libgfortran or anything GPL/LGPL, I think this is fine.

Have you by chance tried this integration when OpenBLAS is present to verify it makes use of it?

(One tiny side note, I think your parent POM still references GPLv2 though the license text does not)

Reply:
> The typical policy is that it's OK to release software that can merely make use of such libraries at runtime (without actually distributing them directly) as long as it doesn't substantially depend on their presence. I believe that dynamic linking in the way you describe is OK - just like having an SPI in JVM code that may be provided by some other GPL code at the user's runtime.

What you describe is exactly how `dev.ludovic.netlib` works. It doesn't substantially depend on OpenBLAS, MKL, or any other native BLAS library to be there as it will fall back to a pure Java implementation otherwise. The transition will be transparent, the feature will be equivalent, only the performance will be affected.

> My main goal is to preserve current behavior.

I fully agree with that, as we don't want to break current behavior nor bring in additional and unwanted dependencies.

> Right now if someone has, say, MKL on their native lib path for the JVM, and built with this alternate profile, it'd be accelerated. If you're saying that still works, but would not require this separate build profile because of the different loading strategy, that's an improvement.

That's exactly how it work with `dev.ludovic.netlib` in JDK16+ today with the implementation based on the Foreign Linker API, and that's how I will want it to work with the JNI-based implementation for JDK8 and JDK11 in the future.

> Have you by chance tried this integration when OpenBLAS is present to verify it makes use of it?

Yes, and it's a lot faster than F2J. The results in https://github.com/apache/spark/pull/32253#issue-619173915 for `native` are with the implementation based on the Foreign Linker API. You can see for `dgemm`, `f2j` is **18-30x slower** than `native` (aka OpenBLAS). I also needed to set `LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu` (on Ubuntu 20.04) for `libblas.so` to be on `ld` path.

To make use of MKL, I only need to set `LD_LIBRARY_PATH=/opt/intel/oneapi/mkl/latest/lib/intel64:/opt/intel/oneapi/compiler/latest/linux/compiler/lib/intel64_lin` and pass `-Ddev.ludovic.netlib.blas.nativeLib=mkl_rt`.

I haven't tried with other BLAS implementations.

Reply:
> (One tiny side note, I think your parent POM still references GPLv2 though the license text does not)

I have fixed that in the latest master. I will push it for Spark right now.

Reply:
OK, it sounds like your solution avoids the small but important nature of the netlib-java shim that made it problematic to include. I think it was libgfortran - the profile didn't itself install BLAS or anything. That is, it wasn't depending on BLAS implementations so much as something in the shim that was linked in. I think.

I don't know this in detail enough to fully evaluate but I believe in your assessment.